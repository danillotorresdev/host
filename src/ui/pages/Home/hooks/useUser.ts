import { useEffect, useState } from "react";
import { User, fetchUser } from "@/services/fetchUser/fetchUser.service";

export const useUser = (username?: string) => {
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);
  const [data, setData] = useState<User | null | undefined>(null);

  useEffect(() => {
    if (!username) return;

    const fetchUserData = async () => {
      setIsLoading(true);
      setError(null);
      try {
        const dataResponse = await fetchUser(username);
        setData(dataResponse);
      } catch (error) {
        setError("Usuário não encontrado");
      } finally {
        setIsLoading(false);
      }
    };

    fetchUserData();
  }, [username]);

  return { data, isLoading, error };
};
