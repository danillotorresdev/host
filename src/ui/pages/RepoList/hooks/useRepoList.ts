import { Repo, fetchRepos } from "@/services/fetchRepos/fetchRepos.service";
import { useEffect, useState } from "react";

export const useRepoList = (userName: string | undefined) => {
  const [repos, setRepos] = useState<Repo[]>([]);
  const [error, setError] = useState<string | null>(null);
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    if (!userName) return;

    const fetchReposData = async () => {
      setIsLoading(true);
      try {
        const repos = await fetchRepos(userName);
        setRepos(repos);
      } catch {
        setError("Erro ao buscar reposit√≥rios. Tente novamente mais tarde.");
      } finally {
        setIsLoading(false);
      }
    };

    fetchReposData();
  }, [userName]);

  return { repos, error, isLoading };
};
