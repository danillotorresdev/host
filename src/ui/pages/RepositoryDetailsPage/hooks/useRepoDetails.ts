import {
  RepoDetail,
  fetchRepoDetails,
} from "@/services/fetchRepoDetails/fetchRepoDetails.service";
import { useEffect, useState } from "react";

type UseRepoDetailsProps = {
  username?: string;
  repositoryName?: string;
};

export const useRepoDetails = ({
  repositoryName,
  username,
}: UseRepoDetailsProps) => {
  const [repositoryData, setRepositoryData] = useState<RepoDetail | null>(null);
  const [isLoading, setIsLoading] = useState(false);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    if (!repositoryName || !username) {
      setError("Repositório não encontrado");
      return;
    }

    const fetchRepoData = async () => {
      try {
        setIsLoading(true);
        const repoData = await fetchRepoDetails(
          `${username}/${repositoryName}`,
        );
        setRepositoryData(repoData);
      } catch (error) {
        setError(
          "Erro ao buscar dados do repositório. Tente novamente mais tarde",
        );
      } finally {
        setIsLoading(false);
      }
    };

    fetchRepoData();
  }, [repositoryName]);

  return {
    repositoryData,
    isLoading,
    error,
  };
};
