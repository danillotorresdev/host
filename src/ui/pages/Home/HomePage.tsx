import * as S from "./HomePage.styles";
import { useState } from "react";
import { useUser } from "@/ui/pages/Home/hooks/useUser";
import { Input } from "@/ui/components/Input/Input";
import { useDebounce } from "@/utils/hooks/useDebounce/useDebounce";
import { CardUserWrapper } from "@/ui/components/CardUserWrapper/CardUserWrapper";

export const HomePage = () => {
  const [username, setUsername] = useState("");
  const userNameDebounced = useDebounce(username, 500);
  const { data, error, isLoading } = useUser(userNameDebounced);
  return (
    <S.HomeWrapper>
      <S.Container>
        <Input
          placeholder="Digite o nome do usuário do GitHub, ex: octocat"
          onChange={(e) => setUsername(e.target.value)}
          label="Usuário do GitHub"
        />
        <CardUserWrapper
          data={data}
          username={userNameDebounced}
          isLoading={isLoading}
          error={error}
        />
      </S.Container>
    </S.HomeWrapper>
  );
};