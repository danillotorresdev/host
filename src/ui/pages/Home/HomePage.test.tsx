import { HomePage } from "./HomePage";
import { describe, expect, it, vi } from "vitest";
import { MemoryRouter } from "react-router-dom"; 
import * as useUser from "./hooks/useUser";
import { render, screen } from "@/utils/settings/tests/utilities";
import { userMock } from "@/services/fetchUser/user.mocks";

describe("HomePage", () => {
  vi.spyOn(useUser, "useUser").mockReturnValue({
    data: userMock,
    isLoading: false,
    error: null,
  });

  it("should render the user data", async () => {
    const { container } = render(
      <MemoryRouter>
        <HomePage />
      </MemoryRouter>,
    );

    expect(container).toMatchSnapshot();
  });

  it("should render the spinner loader", async () => {
    const useUserMock = vi.spyOn(useUser, "useUser");
    useUserMock.mockReturnValue({
      data: null,
      isLoading: true,
      error: null,
    });

    render(
      <MemoryRouter>
        <HomePage />
      </MemoryRouter>,
    );

    expect(screen.getByTestId("spinner-loader")).toBeDefined();
  });

  it("should render the error message", async () => {
    const useUserMock = vi.spyOn(useUser, "useUser");
    useUserMock.mockReturnValue({
      data: null,
      isLoading: false,
      error: "User not found",
    });

    render(
      <MemoryRouter>
        <HomePage />
      </MemoryRouter>,
    );

    expect(screen.getByText("User not found")).toBeDefined();
  });
});
