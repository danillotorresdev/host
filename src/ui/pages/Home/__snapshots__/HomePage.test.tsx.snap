// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HomePage > should render the user data 1`] = `
<div>
  <div
    class="mocked-styled-0 h16k1m1e"
  >
    <div
      class="mocked-styled-2 cf3va6u"
    >
      <div
        class="mocked-styled-4 i19zws4p"
      >
        <label
          class="mocked-styled-5 l5v2eyb"
          for="Usuário do GitHub"
        >
          Usuário do GitHub
        </label>
        <input
          class="mocked-styled-6 i1hv9n31"
          id="Usuário do GitHub"
          placeholder="Digite o nome do usuário do GitHub, ex: octocat"
        />
      </div>
      <div
        class="mocked-styled-10 c1eelvmx"
      >
        <div
          class="mocked-styled-11 p1n234za"
        >
          <div
            class="mocked-styled-15 p1bhgu3m"
          >
            <div
              style="background-image: url(https://); background-repeat: no-repeat; background-position: center center; background-size: cover; filter: blur(20px); transition: filter 0.5s; width: 100%; height: 100%;"
            >
              <img
                alt=""
                src="https://"
                style="display: none;"
              />
            </div>
          </div>
        </div>
        <div
          class="mocked-styled-14 tlp2jul"
        >
          <h2
            class="mocked-styled-12 ngisml4"
          >
            John Doe
          </h2>
          <div
            class="mocked-styled-13 pxniugy"
          >
            <svg
              class="svg svg-followers"
              data-testid="icon-svg"
              style="width: 24px; height: 24px; color: rgb(91, 91, 91);"
              viewBox="0 0 1024 1024"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                xlink:href="/assets/sprite.svg#followers"
              />
            </svg>
            <p>
              Seguidores: 
              10
            </p>
          </div>
          <div
            class="mocked-styled-13 pxniugy"
          >
            <svg
              class="svg svg-followers"
              data-testid="icon-svg"
              style="width: 24px; height: 24px; color: rgb(91, 91, 91);"
              viewBox="0 0 1024 1024"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                xlink:href="/assets/sprite.svg#followers"
              />
            </svg>
            <p>
              Seguindo: 
              10
            </p>
          </div>
          <div
            class="mocked-styled-13 pxniugy"
          >
            <svg
              class="svg svg-email"
              data-testid="icon-svg"
              style="width: 24px; height: 24px; color: rgb(91, 91, 91);"
              viewBox="0 0 1024 1024"
              xmlns="http://www.w3.org/2000/svg"
            >
              <use
                xlink:href="/assets/sprite.svg#email"
              />
            </svg>
            <p>
              johndoe@email.com
            </p>
          </div>
          <div
            class="mocked-styled-13 pxniugy"
          >
            <p>
              My bio
            </p>
          </div>
          <div
            style="width: 200px; margin-top: 1rem;"
          >
            <a
              class="mocked-styled-16 ltiwt9s"
              href="/user//repos"
              style="margin-bottom: 1rem;"
            >
              <svg
                class="svg svg-arrow-foward"
                data-testid="icon-svg"
                style="width: 24px; height: 24px; color: rgb(177, 174, 174);"
                viewBox="0 0 1024 1024"
                xmlns="http://www.w3.org/2000/svg"
              >
                <use
                  xlink:href="/assets/sprite.svg#arrow-foward"
                />
              </svg>
               Repositórios
            </a>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
